In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("weather.csv", index_col="DATE")
df.head()

STATION                                     NAME  ACMH  ACSH  \
DATE                                                                           
1960-01-01  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-02  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-03  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-04  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-05  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   

            AWND  FMTM  FRGT   GAHT  PGTM  PRCP  ...  WT14  WT15  WT16  WT17  \
DATE                                             ...                           
1960-01-01   NaN   NaN   NaN -218.5   NaN  0.00  ...   NaN   NaN   NaN   NaN   
1960-01-02   NaN   NaN   NaN -233.9   NaN  0.38  ...   NaN   NaN   1.0   NaN   
1960-01-03   NaN   NaN   NaN -244.9   NaN  0.01  ...   NaN   NaN   1.0   NaN   
1960-01-04   NaN   NaN   NaN -281.9   NaN  0.00  ...   NaN   NaN   NaN   NaN   
1960-01-05   NaN   NaN   NaN -301.2   NaN  0.00  ...   NaN   NaN   NaN   NaN   

            WT18  WT19  WT21  WT22  WV03  WV20  
DATE                                            
1960-01-01   NaN   NaN   NaN   NaN   NaN   NaN  
1960-01-02   NaN   NaN   NaN   NaN   NaN   NaN  
1960-01-03   1.0   NaN   NaN   NaN   NaN   NaN  
1960-01-04   NaN   NaN   NaN   NaN   NaN   NaN  
1960-01-05   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 50 columns]

In [4]:
# Count null values, % by column 
df.apply(pd.isnull).sum()/df.shape[0]
df.head()

STATION                                     NAME  ACMH  ACSH  \
DATE                                                                           
1960-01-01  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-02  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-03  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-04  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   
1960-01-05  USW00093821  LOUISVILLE INTERNATIONAL AIRPORT, KY US   NaN   NaN   

            AWND  FMTM  FRGT   GAHT  PGTM  PRCP  ...  WT14  WT15  WT16  WT17  \
DATE                                             ...                           
1960-01-01   NaN   NaN   NaN -218.5   NaN  0.00  ...   NaN   NaN   NaN   NaN   
1960-01-02   NaN   NaN   NaN -233.9   NaN  0.38  ...   NaN   NaN   1.0   NaN   
1960-01-03   NaN   NaN   NaN -244.9   NaN  0.01  ...   NaN   NaN   1.0   NaN   
1960-01-04   NaN   NaN   NaN -281.9   NaN  0.00  ...   NaN   NaN   NaN   NaN   
1960-01-05   NaN   NaN   NaN -301.2   NaN  0.00  ...   NaN   NaN   NaN   NaN   

            WT18  WT19  WT21  WT22  WV03  WV20  
DATE                                            
1960-01-01   NaN   NaN   NaN   NaN   NaN   NaN  
1960-01-02   NaN   NaN   NaN   NaN   NaN   NaN  
1960-01-03   1.0   NaN   NaN   NaN   NaN   NaN  
1960-01-04   NaN   NaN   NaN   NaN   NaN   NaN  
1960-01-05   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 50 columns]

In [5]:
# Documentation: keep five core measurements:
# 1. Precipitation (PRCP)
# 2. Snowfall (SNOW)
# 3. Snow depth (SNWD)
# 4. Max temp (TMAX)
# 5. Min temp (TMIN)

core = df[["PRCP", "SNOW", "SNWD", "TMAX", "TMIN"]].copy()
core.head()

PRCP  SNOW  SNWD  TMAX  TMIN
DATE                                    
1960-01-01  0.00   0.0   0.0  41.0  27.0
1960-01-02  0.38   0.0   0.0  48.0  37.0
1960-01-03  0.01   0.0   0.0  43.0  29.0
1960-01-04  0.00   0.0   0.0  33.0  24.0
1960-01-05  0.00   0.0   0.0  33.0  26.0

In [7]:
# Rename columns
core.columns = ["precip", "snow", "snow_depth", "temp_max", "temp_min"]
core.head()

precip  snow  snow_depth  temp_max  temp_min
DATE                                                    
1960-01-01    0.00   0.0         0.0      41.0      27.0
1960-01-02    0.38   0.0         0.0      48.0      37.0
1960-01-03    0.01   0.0         0.0      43.0      29.0
1960-01-04    0.00   0.0         0.0      33.0      24.0
1960-01-05    0.00   0.0         0.0      33.0      26.0

In [11]:
# Percentage of NaN values in columns
core.apply(pd.isnull).sum()/core.shape[0]

precip        0.000350
snow          0.012749
snow_depth    0.012399
temp_max      0.000219
temp_min      0.000219
dtype: float64

In [14]:
core.shape

(22825, 5)

Check the columns to see if they effect the data: Precip, snow, and snow depth. 22,825 entries total from January 1, 1960 to June 28, 2022.

In [13]:
core["precip"].value_counts()

0.00    15032
0.01      710
0.02      501
0.03      349
0.04      271
        ...  
2.40        1
2.24        1
3.31        1
2.80        1
2.92        1
Name: precip, Length: 288, dtype: int64

In [15]:
core["snow"].value_counts()

0.0    21663
0.1      163
0.2      113
0.3       84
0.4       65
       ...  
4.8        1
5.2        1
8.1        1
5.7        1
8.2        1
Name: snow, Length: 78, dtype: int64

In [16]:
core["snow_depth"].value_counts()

0.0     21718
1.0       266
2.0       191
3.0       105
4.0        48
5.0        42
1.2        32
3.1        28
6.0        24
8.0        14
9.0        14
7.0        12
10.0       10
3.9         8
5.1         5
16.0        4
11.0        4
5.9         3
15.0        3
7.1         2
18.0        2
12.0        2
14.0        2
13.0        1
17.0        1
9.8         1
Name: snow_depth, dtype: int64

In [17]:
# Look for NaNs
core[pd.isnull(core["precip"])]

precip  snow  snow_depth  temp_max  temp_min
DATE                                                    
1996-07-02     NaN   NaN         NaN      91.0      71.0
1996-07-03     NaN   NaN         NaN      80.0      63.0
1996-07-04     NaN   NaN         NaN      80.0      59.0
1996-07-05     NaN   NaN         NaN      83.0      61.0
1996-07-06     NaN   NaN         NaN      88.0      61.0
1996-07-07     NaN   NaN         NaN      91.0      70.0
1996-07-08     NaN   NaN         NaN      85.0      70.0
1996-07-09     NaN   NaN         NaN      80.0      66.0

In [18]:
core[pd.isnull(core["snow"])]

precip  snow  snow_depth  temp_max  temp_min
DATE                                                    
1996-02-03    0.00   NaN         NaN      10.0       0.0
1996-02-04    0.00   NaN         NaN      10.0      -5.0
1996-02-05    0.00   NaN         NaN      30.0       1.0
1996-02-06    0.00   NaN         NaN      37.0      16.0
1996-02-07    0.00   NaN         NaN      43.0      33.0
...            ...   ...         ...       ...       ...
2002-10-27    0.00   NaN         NaN      62.0      50.0
2002-10-28    0.32   NaN         NaN      60.0      49.0
2002-10-29    0.48   NaN         NaN      56.0      44.0
2002-10-30    0.01   NaN         NaN      46.0      41.0
2002-10-31    0.01   NaN         NaN      48.0      44.0

[291 rows x 5 columns]

In [19]:
core[pd.isnull(core["snow_depth"])]

precip  snow  snow_depth  temp_max  temp_min
DATE                                                    
1996-02-01    0.03   0.7         NaN      18.0      14.0
1996-02-02    0.00   0.0         NaN      16.0       4.0
1996-02-03    0.00   NaN         NaN      10.0       0.0
1996-02-04    0.00   NaN         NaN      10.0      -5.0
1996-02-05    0.00   NaN         NaN      30.0       1.0
...            ...   ...         ...       ...       ...
2002-10-28    0.32   NaN         NaN      60.0      49.0
2002-10-29    0.48   NaN         NaN      56.0      44.0
2002-10-30    0.01   NaN         NaN      46.0      41.0
2002-10-31    0.01   NaN         NaN      48.0      44.0
2022-06-22    0.00   0.0         NaN     100.0      74.0

[283 rows x 5 columns]

Out of 22,825 entries:
* Precip: 8 NaNs
* Snow: 291 NaNs
* Snow Depth: 283 NaNs

Considering the NaNs are 1.27% or less of total entries, and the other entries are 0s if "none," then it makes sense to assign 0.

In [21]:
core["precip"] = core["precip"].fillna(0)

In [22]:
core["snow"] = core["snow"].fillna(0)

In [23]:
core["snow_depth"] = core["snow_depth"].fillna(0)

In [24]:
# Check that filling NaNs worked
core.apply(pd.isnull).sum()/core.shape[0]

precip        0.000000
snow          0.000000
snow_depth    0.000000
temp_max      0.000219
temp_min      0.000219
dtype: float64

In [25]:
core[pd.isnull(core["temp_max"])]

precip  snow  snow_depth  temp_max  temp_min
DATE                                                    
1996-07-29    0.33   0.0         0.0       NaN       NaN
1996-08-18    0.00   0.0         0.0       NaN       NaN
1996-08-31    0.00   0.0         0.0       NaN       NaN
1996-09-01    0.00   0.0         0.0       NaN       NaN
1996-09-02    0.00   0.0         0.0       NaN       NaN

In [26]:
core[pd.isnull(core["temp_min"])]

precip  snow  snow_depth  temp_max  temp_min
DATE                                                    
1996-07-29    0.33   0.0         0.0       NaN       NaN
1996-08-18    0.00   0.0         0.0       NaN       NaN
1996-08-31    0.00   0.0         0.0       NaN       NaN
1996-09-01    0.00   0.0         0.0       NaN       NaN
1996-09-02    0.00   0.0         0.0       NaN       NaN

In [27]:
# Since temp is unlinkely to be 0, 
# and we only have 5 missing values,
# forward fill

core = core.fillna(method="ffill")

In [28]:
# Check that filling NaNs worked
core.apply(pd.isnull).sum()/core.shape[0]

precip        0.0
snow          0.0
snow_depth    0.0
temp_max      0.0
temp_min      0.0
dtype: float64